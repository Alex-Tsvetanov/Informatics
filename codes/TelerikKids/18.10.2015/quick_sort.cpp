#include <iostream>
#include <random>

using namespace std;

template <typename T>
T* get_pivot (T* a, T* b)
{
	T* masterPivot = a;
	T averrage = 0;
	for (T* it = a ; it != b ; it ++)
		averrage += (*it);
	averrage /= (b - a);
	for (T* it = a ; it != b ; it ++)
		if (abs ((*masterPivot) - averrage) > abs ((*it) - averrage))
			masterPivot = it;

	return masterPivot;
}

template <typename T>
T* fillArr (T* newArr, T* pivot, T* s, T* e)
{
	T* indOfPivot;
	size_t ind = 0;
	for (T* it = s ; it != e ; it ++)
		if ((*it) < (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	indOfPivot = newArr + ind;
	for (T* it = s ; it != e ; it ++)
		if ((*it) >= (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	return indOfPivot;
}

template <typename T>
void copyArr (T* newArr, T* s, T* e)
{
	T* newArrIt = newArr;
	for (T* it = s ; it != e ; it ++, newArrIt ++)
		*it = *newArrIt;
}

template <typename T>
void quick_sort (T* s, T* e)
{
	if (s + 1 == e)
		return ;
	if (s == e)
		return ;
	T* newArr = new T [e - s];
	T* pivot = fillArr (newArr, get_pivot (s, e), s, e);
	//for (T* it = s ; it != e ; it ++)
	//	cout << *it << " ";
	//cout << "\n" << *pivot << "\n----------------------------" << endl;
	quick_sort (newArr, pivot + 1);
	quick_sort (pivot + 1, newArr + (e - s));
	copyArr (newArr, s, e);
	delete[] newArr;
}

int main ()
{
	size_t SIZE;
	cin >> SIZE;
	int* arr = new int [SIZE];
	for (size_t i = 0 ; i < SIZE ; i ++)
		cin >> arr [i];
	quick_sort (arr + 0, arr + SIZE);
	for (size_t i = 0 ; i < SIZE ; i ++)
		cout << arr [i] << " ";
	cout << "\n";
}
