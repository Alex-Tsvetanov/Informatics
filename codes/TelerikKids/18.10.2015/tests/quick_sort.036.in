77
35 22 35 30 6 42 47 83 62 76 42 66 71 43 20 65 24 4 82 41 88 100 27 63 43 23 86 87 7 91 1 67 69 55 68 44 39 29 36 15 53 32 62 30 13 52 33 26 31 52 87 13 93 25 91 29 69 99 40 69 98 26 15 25 56 18 25 65 73 1 8 28 57 70 55 50 44 
