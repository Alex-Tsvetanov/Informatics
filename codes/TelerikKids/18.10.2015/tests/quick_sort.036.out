1 1 4 6 7 8 13 13 15 15 18 20 22 23 24 25 25 25 26 26 27 28 29 29 30 30 31 32 33 35 35 36 39 40 41 42 42 43 43 44 44 47 50 52 52 53 55 55 56 57 62 62 63 65 65 66 67 68 69 69 69 70 71 73 76 82 83 86 87 87 88 91 91 93 98 99 100 
